src/main/scala/com/wavesplatform/history/HistoryWriterImpl.scala

```diff
@@ -0,0 +1,82 @@
+package com.wavesplatform.history
+
+import com.google.common.cache.{CacheBuilder, CacheLoader}
+import com.google.common.util.concurrent.UncheckedExecutionException
+import scorex.account.Account
+import scorex.block.Block
+import scorex.block.Block.BlockId
+import scorex.transaction.History.BlockchainScore
+import scorex.transaction.ValidationError.CustomError
+import scorex.transaction.{History, HistoryWriter, ValidationError}
+import scorex.utils.ScorexLogging
+
+class HistoryWriterImpl(storage: HistoryStorage) extends History with HistoryWriter with ScorexLogging {
```

Comment:
Three different approaches to error handling are used in a single class: returning Either, returning an option with no clue and just throwing an exception. This doesn't look neat to me.